### PR TITLE
feat: use slug format for atlas short names in urls and s3 keys (#1142)

### LINKED ISSUE #1142: Atlas short name validation and slugification
## Summary

Restrict atlas `shortName` to alphabetic characters and spaces only, and introduce a deterministic slug (kebab-case) derived from the short name for use in S3 keys, concept table storage, and URL-based atlas resolution.

## Motivation

Atlas short names currently accept any string with no character restrictions. Names like "Ocular Outflow Segment" (with spaces) cannot be used reliably in S3 keys or URLs. By restricting input to `[a-zA-Z ]` and slugifying for machine contexts, we get:

- **Unambiguous S3 keys**: `eye/ocular-outflow-segment-v1/source-datasets/file.h5ad`
- **Clean URLs**: `/api/atlases/ocular-outflow-segment_v1.0`
- **Bijective mapping**: The restricted character set guarantees no two valid short names produce the same slug
- **No digits in slugs**: This means `-v\d+` in S3 keys always unambiguously marks the version boundary

### Input Validation

- **Allowed characters**: `[a-zA-Z ]` (upper/lowercase alpha and space)
- **Regex**: `/^[a-zA-Z]+( [a-zA-Z]+)*$/`
  - No leading/trailing spaces
  - No consecutive spaces
  - No hyphens, underscores, digits, or special characters
- **Examples**: `"Gut"`, `"Retina"`, `"Ocular Outflow Segment"`

### Slug Function

```typescript
function slugifyShortName(shortName: string): string {
  return shortName.toLowerCase().replace(/ /g, "-");
}
```

No external library needed. The restricted input makes this trivial and deterministic.

### Storage

| Context | Format | Example |
|---|---|---|
| `atlases.overview.shortName` (JSONB) | Display form (as entered) | `"Ocular Outflow Segment"` |
| `concepts.atlas_short_name` | Slug | `"ocular-outflow-segment"` |
| S3 key | Slug + version | `eye/ocular-outflow-segment-v1/...` |
| URL parameter | Slug + version | `ocular-outflow-segment_v1.0` |

### Matching

Current:
```sql
lower(a.overview->>'shortName') = c.atlas_short_name
```

After (conceptually the same for single-word names, now correct for multi-word names too):
- The slug stored in `concepts.atlas_short_name` is produced by the slugify function
- Matching from S3 keys: the parsed base name IS already the slug (since S3 keys use the slug)
- Matching from atlas overview: apply slugify to `overview->>'shortName'` or use `lower()` + `replace(' ', '-')`

---

### Phase 1: Add Slug Utility and Validation

**Files to create/modify:**

1. **Create `app/utils/slug.ts`** (or add to `app/utils/atlases.ts`)
   - `slugifyShortName(shortName: string): string` — lowercase + spaces to hyphens
   - `ATLAS_SHORT_NAME_REGEX = /^[a-zA-Z]+( [a-zA-Z]+)*$/`
   - Optional: `isValidAtlasShortName(name: string): boolean`

2. **Update `app/apis/catalog/hca-atlas-tracker/common/schema.ts`** (line 97)
   - Add `.matches()` validation to the `shortName` field:
     ```typescript
     shortName: string()
       .default("")
       .required("Short name is required")
       .matches(
         /^[a-zA-Z]+( [a-zA-Z]+)*$/,
         "Short name must contain only letters and single spaces"
       ),
     ```

### Phase 2: Update Concept Table Logic

**Files to modify:**

3. **Update `app/utils/files.ts`** (line 138)
   - Currently: `atlasShortName: atlasBaseName.toLowerCase()`
   - The S3 key already contains the slug, so `toLowerCase()` is sufficient (the slug has no spaces). No change strictly needed here since S3 keys will use slugs, but import and use `slugifyShortName` for clarity/consistency:
     ```typescript
     atlasShortName: slugifyShortName(atlasBaseName),
     // For slug input this is equivalent to toLowerCase(), but explicit
     ```

4. **Update `app/data/concepts.ts`** (line 106)
   - Current matching:
     ```sql
     AND lower(a.overview->>'shortName') = c.atlas_short_name
     ```
   - Update to:
     ```sql
     AND lower(replace(a.overview->>'shortName', ' ', '-')) = c.atlas_short_name
     ```
   - This converts the display name to a slug in SQL for matching

5. **Update `app/data/atlases.ts`** (lines 20-22)
   - `getAtlasIdByNameAndVersion()` currently does:
     ```typescript
     const lowerShortName = shortName.toLowerCase();
     // ... WHERE LOWER(overview->>'shortName') = $1
     ```
   - Update to use slug comparison:
     ```typescript
     const slugName = slugifyShortName(shortName);
     // ... WHERE lower(replace(overview->>'shortName', ' ', '-')) = $1
     ```
   - Note: The `shortName` parameter here comes from URL parsing (`parseUrlAtlasName`), which already receives the slug from the URL. So this is slug-to-slug matching.

6. **Update `app/services/concepts.ts`**
   - Verify `getOrCreateConceptId()` receives the slug (it does — comes from S3 key parsing). No logic change needed, but audit the flow.

7. **Update `app/utils/atlases.ts`** (line 17-26, `parseUrlAtlasName`)
   - The regex `/^(.+)_v([1-9]\d*)\.(0|[1-9]\d*)$/` already handles slugs with hyphens correctly (greedy `.+` captures everything before `_v`)
   - No regex change needed, but the returned `shortName` is now a slug — document this

8. **Update `db_scripts/generate-test-validation-results.ts`** (lines 248-249)
   - Update any `LOWER(overview->>'shortName')` usage to use slug-based comparison

### Phase 3: Database Migration

**Create new migration:**

9. **Create migration `migrations/{timestamp}_slugify_atlas_short_names.ts`**

   This migration must:

   a. **Update `concepts.atlas_short_name`** — convert any existing values that have spaces to slugs:
      ```sql
      UPDATE hat.concepts
      SET atlas_short_name = lower(replace(atlas_short_name, ' ', '-'));
      ```
      (Current data is already lowercase with hyphens, so this is a no-op for existing data but protects against any edge cases.)

   b. **Optionally add a CHECK constraint** on `concepts.atlas_short_name` to enforce slug format at the DB level:
      ```sql
      ALTER TABLE hat.concepts
      ADD CONSTRAINT chk_atlas_short_name_slug
      CHECK (atlas_short_name ~ '^[a-z]+(-[a-z]+)*$');
      ```

   c. **Down migration**: Remove the CHECK constraint

10. **Run `npm run dump-schema`** to update `schema.sql`

### Phase 4: Update Test Data

**Files to modify:**

11. **Update `testing/constants.ts`**
    - Convert all hyphenated `shortName` values to space-separated display names:
      | Current | New |
      |---|---|
      | `"test-draft"` | `"Test Draft"` |
      | `"test-public"` | `"Test Public"` |
      | `"test-with-il"` | `"Test With Il"` |
      | `"test-with-misc-source-studies"` | `"Test With Misc Source Studies"` |
      | `"test-with-misc-source-studies-b"` | `"Test With Misc Source Studies B"` |
      | `"test-with-misc-source-studies-c"` | `"Test With Misc Source Studies C"` |
      | `"test-with-source-study-validations-b"` | `"Test With Source Study Validations B"` |
      | `"test-public-bar"` | `"Test Public Bar"` |
      | `"test-with-metadata-correctness"` | `"Test With Metadata Correctness"` |
      | `"test-public-baz"` | `"Test Public Baz"` |
      | `"test-with-cap-id"` | `"Test With Cap Id"` |
      | `"test-with-entry-sheet-validations-a"` | `"Test With Entry Sheet Validations A"` |
      | `"test-with-entry-sheet-validations-b"` | `"Test With Entry Sheet Validations B"` |
      | `"test-with-non-shared-entry-sheet-validations"` | `"Test With Non Shared Entry Sheet Validations"` |
      | `"heatmap-test-atlas"` | `"Heatmap Test Atlas"` |
      | `"test-with-non-latest-metadata-entities"` | `"Test With Non Latest Metadata Entities"` |
      | `"test-published"` | `"Test Published"` |
      | `"test-with-linked-publish-statuses"` | `"Test With Linked Publish Statuses"` |
    - Single-word names like `"Gut"` and `"Retina"` remain unchanged

12. **Update `testing/db-utils.ts`** (line 280)
    - Currently: `atlas_short_name: atlas.shortName.toLowerCase()`
    - Change to: `atlas_short_name: slugifyShortName(atlas.shortName)`

13. **Update `testing/utils.ts`** (line 333)
    - S3 key construction: `${atlas.network}/${atlas.shortName}-v${atlas.generation}-${atlas.revision}/${folderName}/${file.fileName}`
    - Change to use slug: `${atlas.network}/${slugifyShortName(atlas.shortName)}-v${atlas.generation}-${atlas.revision}/${folderName}/${file.fileName}`

14. **Update `testing/sns-testing.ts`** (lines 318, 328, 338, 348)
    - `shortName` values like `"test-short-name-a"`, `"test-short-name-b"`, `"test-published-atlas"` → convert to display form
    - S3 keys in test fixtures that embed the atlas name must use the slug form

15. **Update `__tests__/api-sns-with-s3-event.test.ts`**
    - All `atlas_short_name` assertions (lines 1552-1862) — values should be slugs derived from the updated display names
    - S3 key strings in test data must use slug format
    - Assertions comparing concept `atlas_short_name` to expected values

16. **Update `__tests__/api-atlases-create.test.ts`** and `__tests__/api-atlases-id.test.ts`**
    - Add test cases for validation rejection of invalid short names (hyphens, digits, underscores, special chars)
    - Add test cases for valid short names with spaces

### Phase 5: Verify and Run Tests

17. **Run linting**: `npm run lint`
18. **Run format check**: `npm run check-format`
19. **Run full test suite**: `npm run test`
20. **Fix any failures from the short name format change**

---

### New files
- `app/utils/slug.ts` (or addition to `app/utils/atlases.ts`)
- `migrations/{timestamp}_slugify_atlas_short_names.ts`

### Modified files — Application code
- `app/apis/catalog/hca-atlas-tracker/common/schema.ts` — validation regex
- `app/utils/files.ts` — S3 key parsing (use slugify for clarity)
- `app/data/concepts.ts` — SQL matching logic
- `app/data/atlases.ts` — atlas name resolution query
- `db_scripts/generate-test-validation-results.ts` — test data generation

### Modified files — Tests
- `testing/constants.ts` — all shortName values
- `testing/db-utils.ts` — concept creation
- `testing/utils.ts` — S3 key construction, assertions
- `testing/sns-testing.ts` — SNS test fixtures
- `__tests__/api-sns-with-s3-event.test.ts` — concept assertions
- `__tests__/api-atlases-create.test.ts` — new validation tests
- `__tests__/api-atlases-id.test.ts` — new validation tests

### Modified files — Schema
- `schema.sql` — updated via `npm run dump-schema`

---

## Edge Cases and Considerations

1. **Existing production data**: Any atlas with a `shortName` containing hyphens, digits, or special characters must be manually corrected before deploying the validation change. Coordinate with content admins.

2. **S3 key parsing with hyphens in slug**: The regex `/^(.+)-v(\d+(?:-\d+)*)$/` works correctly because `.+` is greedy and `-v` followed by digits is unambiguous. Since slugs cannot contain digits (alpha-only input), `-v\d+` always marks the version boundary.

3. **URL parsing**: `parseUrlAtlasName` uses `_v` as the version separator, which is distinct from hyphens in the slug. `ocular-outflow-segment_v1.0` parses correctly.

4. **Case sensitivity**: Display names preserve case (`"Ocular Outflow Segment"`), slugs are always lowercase (`"ocular-outflow-segment"`). The slug function handles this.

5. **Uniqueness**: Two atlases cannot have short names that differ only in case (e.g., `"Gut"` vs `"gut"`) because they'd produce the same slug. This is already the case with the current `lower()` matching, so no behavioral change.

6. **Backwards compatibility of API URLs**: URLs like `/api/atlases/gut_v1.0` continue to work. Multi-word names now use hyphens: `/api/atlases/ocular-outflow-segment_v1.0`.

### PR DESCRIPTION
**Pull request by hunterckx**
Closes #1142

Notes:
- Most of the test atlases had to be updated to have their short names in the letters-and-spaces format
- I renamed some things (mostly around the newly-named `parseAtlasNameUrlSlug`) to be more clear that they relate to slugs rather than title-format short names

### CODE REVIEW

File: app/utils/files.ts [hunk 1]
```diff
@@ -71,18 +75,15 @@ function determineFileType(s3TypeFolder: string): FILE_TYPE {
...
- * parseS3AtlasName('gut-v1') // Returns: { atlasBaseName: 'gut', s3Version: '1' }
- * parseS3AtlasName('retina-v1-1') // Returns: { atlasBaseName: 'retina', s3Version: '1-1' }
+ * parseS3AtlasName('gut-v1') // Returns: { shortNameSlug: 'gut', generation: 1, revision: 0 }
+ * parseS3AtlasName('retina-v1-1') // Returns: { shortNameSlug: 'retina', generation: 1, revision: 1 }
  */
-function parseS3AtlasName(s3AtlasName: string): {
-  atlasBaseName: string;
-  s3Version: string;
-} {
+function parseS3AtlasName(s3AtlasName: string): AtlasSlugNameAndVersion {
```
Comment (hunterckx) [inline app/utils/files.ts]: Updated the return value for this to be a bit more consistent with `parseAtlasNameUrlSlug`

**Review by NoopDog — APPROVED** [review]
